presto-verifier/src/main/java/com/facebook/presto/verifier/framework/DataVerification.java

```diff
@@ -91,23 +102,54 @@ public DataMatchResult verify(
-
-        controlChecksumQueryContext.setChecksumQuery(formatSql(controlChecksumQuery));
         testChecksumQueryContext.setChecksumQuery(formatSql(testChecksumQuery));
 
-        QueryResult<ChecksumResult> controlChecksum = callAndConsume(
-                () -> getHelperAction().execute(controlChecksumQuery, CONTROL_CHECKSUM, ChecksumResult::fromResultSet),
-                stats -> stats.getQueryStats().map(QueryStats::getQueryId).ifPresent(controlChecksumQueryContext::setChecksumQueryId));
+        List<Column> controlColumns = null;
+        ChecksumResult controlChecksumResult = null;
+
+        if (!skipControl) {
+            controlColumns = getColumns(getHelperAction(), typeManager, control.getObjectName());
+            Query controlChecksumQuery = checksumValidator.generateChecksumQuery(control.getObjectName(), controlColumns);
+            controlChecksumQueryContext.setChecksumQuery(formatSql(controlChecksumQuery));
+
+            QueryResult<ChecksumResult> controlChecksum = callAndConsume(
+                    () -> getHelperAction().execute(controlChecksumQuery, CONTROL_CHECKSUM, ChecksumResult::fromResultSet),
+                    stats -> stats.getQueryStats().map(QueryStats::getQueryId).ifPresent(controlChecksumQueryContext::setChecksumQueryId));
+            controlChecksumResult = getOnlyElement(controlChecksum.getResults());
+        }
+        else if (!saveSnapshot) {
+            controlColumns = testColumns;
+            String key = format(VERIFIER_SNAPSHOT_KEY_PATTERN, getSourceQuery().getSuite(), getSourceQuery().getName(), isExplain);
+            SnapshotQuery snapshotQuery = snapshotQueries.get(key);
+            if (snapshotQuery != null) {
+                String snapshotJson = snapshotQuery.getSnapshot();
+                controlChecksumResult = ChecksumResult.fromJson(snapshotJson);
+            }
+            else {
+                return new DataMatchResult(SNAPSHOT_DOES_NOT_EXIST, Optional.empty(), OptionalLong.empty(), OptionalLong.empty(), Collections.emptyList());
+            }
+        }
+
         QueryResult<ChecksumResult> testChecksum = callAndConsume(
                 () -> getHelperAction().execute(testChecksumQuery, TEST_CHECKSUM, ChecksumResult::fromResultSet),
                 stats -> stats.getQueryStats().map(QueryStats::getQueryId).ifPresent(testChecksumQueryContext::setChecksumQueryId));
+        ChecksumResult testChecksumResult = getOnlyElement(testChecksum.getResults());
+
+        if (saveSnapshot) {
+            String snapshot = ChecksumResult.toJson(Optional.of(testChecksumResult));
+
+            snapshotQueryConsumer.accept(new SnapshotQuery(getSourceQuery().getSuite(), getSourceQuery().getName(), isExplain, snapshot));
+            return new DataMatchResult(
+                    MATCH,
+                    Optional.empty(),
+                    OptionalLong.empty(),
+                    OptionalLong.empty(),
+                    ImmutableList.of());
```

Comment:
curious why we are returning a MATCH here.?

Should I always have to run snapshot first and then use the query bank. Can't I save the results along with a regular query run.

If saving snapshots can't be combined with a regular run (I think we should support it, no reason not to)
then this check `if (saveSnapshot) {` should be the first check in this method. We want to signal to the reader that we are not doing any Data verification here, just saving the results. (In fact, you have followed the pattern of exiting early in ExplainVerification :) )

Reply:
Returning a match to make it as succeeded in the stats, which looks better.

If we support saving with regular run, the stats may look confusing. What if the regular run succeeds but the saving fails?

And yes, ` if (saveSnapshot)` should be the first check and it was the first check originally. But some existing tests depend on the order when checksums are calculated, `controlChecksum `must be calculated before `testChecksum`.

Reply:
If we use controlChecksum as I recommended in https://github.com/prestodb/presto/pull/19253#discussion_r1275623295 your requirements of controlChecksum has to be computed before testChecksum will be satisfied.

Also super curious which tests fail if the order is inverted.

Reply:
Current design sets `skipControl = true` when running in `query-bank` mode. So it is not possible to save control checksum.

Reply:
I think this is where the current design is not clean.

Let us draw up an analogy with a regular bank.
A regular bank accepts deposits and loans money.
Similarly our query bank can save snapshots and give out snapshots.

The saving snapshot part/deposit should be always based on control since that is the correct or vetted version of presto. It is confusing (and wrong IMHO) to say that saving snapshots is going to be based off a test version.

While running the test in query bank mode, the saved snapshot from control is fetched and compared with the values obtained from the test version.

So the **tweak** in design is to understand the **intent** of the test run in query bank mode. Is it a deposit or a withdrawal?.
If you can incorporate that I think the design will be a lot cleaner.

Reply:
I'll discuss with Eric.

Reply:
Some reasons we decided to use test result:

- We may want to run only one set of the tests in the future, that should be test, not control. The code for control can be deleted to further simplify the logic.
- No need to run both tests in `query-bank` mode, since we already have existing code to `skip-control`, why not utilize it.
- `save-snapshot` is one time thing. After a successful regular run, both control and test are equally good, saving test result is Ok.
- In case there is an expected change in the result, it is in the test, not control. In this case, test result should be saved.

Reply:
Discussed with Eric, I'm changing it to save control checksum.

Reply:
@ajaygeorge changed to save control as snapshot. Thanks for reviewing.